src/zed_wrapper_node.cpp

```diff
@@ -53,57 +54,119 @@
-    left_im.header.frame_id = left_frame_id;
-    left_im.header.stamp = t;
-    pub_left.publish(left_im.toImageMsg());
+/* \brief Publish a cv::Mat depth image with a ros Publisher
+ * \param depth : the depth image to publish
+ * \param pub_depth : the publisher object to use
+ * \param depth_frame_id : the id of the reference frame of the depth image
+ * \param t : the ros::Time to stamp the depth image
+ */
+void publishDepth(cv::Mat depth, image_transport::Publisher &pub_depth, string depth_frame_id, ros::Time t) {
+    cv_bridge::CvImage depth_im;
+    depth_im.image = depth;
+    depth_im.encoding = sensor_msgs::image_encodings::TYPE_16UC1;
+    depth_im.header.frame_id = depth_frame_id;
+    depth_im.header.stamp = t;
+    pub_depth.publish(depth_im.toImageMsg());
+}
 
-    // Publish second image
-    cv_bridge::CvImage second_im;
-    second_im.image = second;
-    second_im.encoding = (computeDepth) ? sensor_msgs::image_encodings::TYPE_16UC1 : sensor_msgs::image_encodings::BGR8;
-    second_im.header.frame_id = second_frame_id;
-    second_im.header.stamp = t;
-    pub_second.publish(second_im.toImageMsg());
 
+/* \brief Publish a pointCloud with a ros Publisher
+ * \param p_could : the float pointer to point cloud datas
+ * \param width : the width of the point cloud
+ * \param height : the height of the point cloud
+ * \param pub_cloud : the publisher object to use
+ * \param cloud_frame_id : the id of the reference frame of the point cloud
+ * \param t : the ros::Time to stamp the point cloud
+ */
+void publishPointCloud(float* p_cloud, int width, int height, ros::Publisher &pub_cloud, string cloud_frame_id, ros::Time t) {
+    pcl::PointCloud<pcl::PointXYZRGB> point_cloud;
+    point_cloud.width = width;
+    point_cloud.height = height;
+    int size = width*height;
+    point_cloud.points.resize(size);
+    int index4 = 0;
+    float color;
+    for (int i = 0; i < size; i++) {
+        if(p_cloud[index4+2] < 0) { // Check if it's an unvalid point, the depth is lower than 0 
+            index4 += 4;
+            continue; 
+        }
+        pcl::PointXYZRGB point;
+        point.y = -p_cloud[index4++] * 0.001;
+        point.z = -p_cloud[index4++] * 0.001;
+        point.x = p_cloud[index4++] * 0.001;
+        color = p_cloud[index4++];
+        uint32_t color_uint = *(uint32_t*) & color; // Convert the color
+        unsigned char* color_uchar = (unsigned char*) &color_uint;
+        color_uint = ((uint32_t) color_uchar[0] << 16 | (uint32_t) color_uchar[1] << 8 | (uint32_t) color_uchar[2]);
+        point.rgb = *reinterpret_cast<float*> (&color_uint);
+        point_cloud.points[i] = point;
+    }
+    sensor_msgs::PointCloud2 output;
+    pcl::toROSMsg(point_cloud, output); // Convert the point cloud to a ROS message
+    output.header.frame_id = cloud_frame_id;// Set the header values of the ROS message
+    output.header.stamp = t;
+    pub_cloud.publish(output);
```

Comment:
Pcl will set the height to be 1, if not initialized. In this case, the pointcloud is structured, so IMO it makes sense for the clouds dimensions to match the image dimensions